Services/Object/Icon/classes/class.ilObjectCustomIconFactory.php

```diff
@@ -78,4 +91,36 @@ public function getByObjId($objId, $objType = '')
 			$objId
 		);
 	}
+
+	/**
+	 * Get custom icon presenter
+	 *
+	 * @param $objId
+	 * @param $objType
+	 *
+	 * @return \ilObjectCustomIconPresenter
+	 */
+	public function getPresenterByObjId($objId, $objType)
+	{
+		if (0 === strlen($objType)) {
+			$objType = $this->objectCache->lookupType($objId);
+		}
+
+		$presenter = null;
+		switch($objType)
+		{
+			case 'catr':
+			case 'grpr':
+			case 'crsr':
+				$target_obj_id = \ilObjectReferenceCustomIconPresenter::lookupTargetId($objId, $this->dbInterface);
```

Comment:
Do you think we can get rid of the static method call here?

We already have a lookup method here: `ilContainerReference::_lookupTargetId($a_obj_id);`

I suggest to inject some class as a dependency to the constructor of this factory, which internally wraps this lookup and can be used here.

Reply:
@mjansenDatabay You are absolutely right, the method _lookupTargetId should be reused. 
Regarding your suggestion to avoid the static method, I see three possibilities: 

- as you said add the "ContainerReference" as dependency and determine the target ref_id. But after this is done, an instance of ilObjectCustomIconImpl needs to be created and this would require to add the factory itself as dependency for the "ilObjectReferenceCustomIconPresenter". => something I like to avoid

- keep the static method to look up the target id and then create a ilObjectCustomIconPresenter without further dependencies.

- get rid of the static method and directly call ilContainerReference::_lookupTargetId in the factory for container references.

Reply:
@smeyer-ilias As you already wrote the dependency could be passed to the factory constructor AND delegated to the constructor of `ilObjectReferenceCustomIconPresenter` if an instance is created.

If you want to avoid this, I prefer option 3.